### PR TITLE
HPE Aruba clearpass v1

### PR DESCRIPTION
**Pull request by abaumgarten**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/6683

## Description
A new integration - HPE Arube Clearpass
includes:
8 commands:

1. **aruba-clearpass-endpoints-list**
2. **aruba-clearpass-endpoint-update**
3. **aruba-clearpass-attributes-list**
4. **aruba-clearpass-attribute-create**
5. **aruba-clearpass-attribute-update**
6. **aruba-clearpass-attribute-delete**
7. **aruba-clearpass-active-sessions-list**
8. **aruba-clearpass-active-session-disconnect**
test-module
TPB
UTs
README

note:
The following commands were not tested (as we don't have active session in Aruba, we asked the vendor to set us some):
**aruba-clearpass-active-sessions-list**
**aruba-clearpass-active-session-disconnect**

## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 5.5.0
- [x] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [x] Documentation

### CODE REVIEW

File: Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py [hunk 1]
```diff
@@ -0,0 +1,472 @@
...
+    endpoints_filter = json.dumps(endpoints_filter)  # the API requires the value of 'filter' to be a json object.
+    params = {"filter": endpoints_filter, "limit": limit, "offset": offset}
+
+    res = client.prepare_request(method='GET', params=params, url_suffix='endpoint')
+    readable_output, outputs = parse_items_response(res, endpoints_response_to_dict)
+    human_readable = tableToMarkdown('HPE Aruba Clearpass endpoints', readable_output, removeNull=True)
+
+    return CommandResults(
+        readable_output=human_readable,
+        outputs_prefix='HPEArubaClearpass.endpoints.list',
```
Comment (guykeller) [inline Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py:182]: why `.list`?
same everywhere else

File: Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py [hunk 2]
```diff
@@ -0,0 +1,472 @@
...
+        outputs_key_field='id',
+        outputs=outputs,
+    )
+
+
+def endpoints_response_to_dict(response: dict) -> dict:
```
Comment (guykeller) [inline Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py:188]: but it is already a dict

File: Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py [hunk 3]
```diff
@@ -0,0 +1,472 @@
...
+    attributes_values = {}
+    for attribute in attributes:  # converting the given list of attributes pairs to a a dict
+        attributes_values.update(attribute)
+
+    request_body = {}
+    request_body.update({"status": status}) if status else None
+    request_body.update({"mac_address": mac_address}) if mac_address else None
+    request_body.update({"description": description}) if description else None
+    request_body.update({"device_insight_tags": device_insight_tags}) if device_insight_tags else None
+    request_body.update({"attributes": attributes_values}) if attributes else None
```
Comment (guykeller) [inline Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py:218]: as discussed

File: Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py [hunk 4]
```diff
@@ -0,0 +1,472 @@
...
+
+
+def create_new_attribute_body(args: Dict[str, Any]):
+    """
+    Creates a new attribute body for creating and updating an attribute.
+    """
+    name = args.get('name')
+    entity_name = args.get('entity_name')
+    data_type = args.get('data_type')
+    mandatory = args.get('mandatory', False)
```
Comment (guykeller) [inline Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py:304]: Is it a string or a boolean?
Same goes for other places

File: Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py [hunk 5]
```diff
@@ -0,0 +1,472 @@
...
+    url_suffix = f"/session/{session_id}/disconnect"
+    body = {"id": session_id, "confirm_disconnect": True}
+
+    res = client.prepare_request(method='POST', params={}, url_suffix=url_suffix, body=body)
+    outputs = {"Error_code": res.get('error'), "Response_message": res.get('message')}
+    human_readable = tableToMarkdown('HPE Aruba Clearpass Disconnect active session', outputs, removeNull=True)
+
+    return CommandResults(
+        readable_output=human_readable,
+        outputs_prefix='HPEArubaClearpass.sessions.disconnect',
```
Comment (guykeller) [inline Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py:414]: Why under disconnect? Shouldn't we just update the sessions path that the session has disconnected?

File: Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass_test.py [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+    - Ensures that if access token is expired, a new one is generated.
+    """
+    client = create_client()
+    mocker.patch.object(HPEArubaClearpass, "get_integration_context", return_value=integration_context)
+    mocker.patch.object(client, "generate_new_access_token", return_value=CLIENT_AUTH)
+    client.login()
+    assert client.access_token == expected_token
+
+
+def test_get_endpoints_list_command(mocker):
```
Comment (guykeller) [inline Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass_test.py:61]: General note - I'm more interested in testing Pythonic logic than the commands. Testing helper methods is more indicative and helpful IMO

File: Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/README.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+This README contains the full documentation for your integration.
```
Comment (guykeller) [inline Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/README.md:1]: let's update this, and add command examples too

File: Packs/HPEArubaClearpass/pack_metadata.json [hunk 8]
```diff
@@ -0,0 +1,15 @@
+{
+    "name": "HPE Aruba Clearpass",
+    "description": "change it",
```
Comment (guykeller) [inline Packs/HPEArubaClearpass/pack_metadata.json:3]: change it indeed

File: Packs/HPEArubaClearpass/pack_metadata.json [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+    "currentVersion": "1.0.0",
+    "author": "Cortex XSOAR",
+    "url": "https://www.paloaltonetworks.com/cortex",
+    "email": "",
+    "categories": [
+        "Network Security"
+    ],
+    "tags": [],
+    "useCases": [],
+    "keywords": []
```
Comment (guykeller) [inline Packs/HPEArubaClearpass/pack_metadata.json:14]: let's add

**Review by guykeller — CHANGES_REQUESTED** [review]
Let's do the self CR we discussed

File: Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+    return message
+
+
+def parse_items_response(response: dict, active_sessions_parsing=None):  # type:ignore
+    items_list = response.get('_embedded', {}).get('items')
+    human_readable = []
+    if items_list:
+        for item in items_list:
+            if item.get('_links'):
+                del item['_links']  # useless information, won't be displayed
```
Comment (guykeller) [inline Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py:137]: we have a function for that, don't we?

**Review by guykeller — APPROVED** [review]
last few notes.
Also, let's schedule a demo with Anar, Guy and Yuval before merging

File: Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py [hunk 11]
```diff
@@ -0,0 +1,469 @@
...
+
+    return CommandResults(
+        readable_output=human_readable,
+        outputs_prefix='HPEArubaClearpass.endpoints',
+        outputs_key_field='id',
+        outputs=outputs,
+    )
+
+
+def delete_redundant_data(res: Dict[str, Any]):
```
[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py:201]: Added

File: Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py [hunk 12]
```diff
@@ -0,0 +1,469 @@
...
+    try:
+        del res['_links']
+    except KeyError:
+        pass
```
[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py]: Done

File: Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py (hunk 10 above)
Comment (abaumgarten) [reply Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py]: You right, done.

File: Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py [hunk 13]
```diff
@@ -0,0 +1,469 @@
...
+    attribute_filter.update({"id": attribute_id}) if attribute_id else None
+    attribute_filter.update({"name": name}) if name else None
+    attribute_filter.update({"entity_name": entity_name}) if entity_name else None
+    return json.dumps(attribute_filter)  # the API requires the value of 'filter' to be a json object.
+
+
+def check_api_limitation_on_specific_data_types(args: Dict[str, Any]):
+    """ Checks if the attribute data_type match the api limitations like:
+    1. allow_multiple is available only when data_type is String.
+    1. allowed_value is available only when data_type is List.
```
[the remark this answers is not in the document]
Comment (abaumgarten) [reply Packs/HPEArubaClearpass/Integrations/HPEArubaClearpass/HPEArubaClearpass.py]: The api docs are private (need user and password). I added the suffix of the link.

**Comment by abaumgarten** [comment]
@kirbles19 assign you as this is a new integration that requires docs review. thanks!

**Comment by ShirleyDenkberg** [comment]
@guykeller Doc review completed.

**Review by guykeller — APPROVED** [review]
Pending customer verification

**Pull request merged**